File: Android/SayHi/app/src/main/java/de/oklemenz/sayhi/model/Category.java
package de.oklemenz.sayhi.model;

import android.graphics.Bitmap;
import android.graphics.Color;
import android.text.TextUtils;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.Serializable;

import de.oklemenz.sayhi.service.Cache;
import de.oklemenz.sayhi.service.IconService;
import de.oklemenz.sayhi.service.Utilities;

/**
 * Created by Oliver Klemenz on 31.10.16.
 */

public class Category implements Serializable {

    public static String CategoryStagedIcon = "staged";
    public static String CategoryStagedColor = "#999999";

    public static Category CategoryFavorite = new Category(1, "favorite");
    public static Category CategorySearch = new Category(2, "search");
    public static Category CategoryOwn = new Category(3, "own");
    public static Category CategoryStaged = new Category(4, "staged");
    public static Category CategoryMore = new Category(5, "more");

    public String key = "";
    public String langCode = "";
    private String name = "";

    public void setName(String name) {
        this.name = name;
        this.search = Utilities.searchNormalized(name, langCode);
    }

    public String getName() {
        return this.name;
    }

    private String color = "";

    public void setColor(String color) {
        this.color = color;
        if (!TextUtils.isEmpty(color)) {
            this.bgColor = Color.parseColor(color);
            this.textColor = Utilities.isColorLight(this.bgColor) ? Color.BLACK : Color.WHITE;
        } else {
            this.bgColor = Color.TRANSPARENT;
            this.textColor = Color.WHITE;
        }
    }

    public String getColor() {
        return color;
    }

    private String icon = "";

    public void setIcon(String icon) {
        this.icon = icon;
    }

    public String getIcon() {
        return this.icon;
    }

    public long order = 0;

    public String primaryLangKey = "";
    public String refKey = "";
    private String refPrimaryLangKey;

    public void setRefPrimaryLangKey(String refPrimaryLangKey) {
        this.refPrimaryLangKey = refPrimaryLangKey;
    }

    public String getRefPrimaryLangKey() {
        return this.refPrimaryLangKey != null ? this.refPrimaryLangKey : this.refKey;
    }

    public int bgColor = Color.WHITE;
    public int textColor = Color.BLACK;
    private transient Bitmap iconImage;

    public Bitmap getIconImage() {
        if (!TextUtils.isEmpty(icon)) {
            if (iconImage == null) {
                iconImage = IconService.getInstance().icon(icon);
            }
        }
        return iconImage;
    }

    public String search = "";
    public boolean favorite = false;
    public boolean selected = false;
    public boolean stage = false;
    public int mark = 0;

    public String space = "";

    public Category(String key, String langCode, String name, String color, String icon, long order, String primaryLangKey) {
        this(key, langCode, name, color, icon, order, primaryLangKey, null, null);
    }

    public Category(String key, String langCode, String name, String color, String icon, long order, String primaryLangKey, String refKey, String refPrimaryLangKey) {
        this(key, langCode, name, color, icon, order, primaryLangKey, refKey, refPrimaryLangKey, false);
    }

    public Category(String key, String langCode, String name, String color, String icon, long order, String primaryLangKey, String refKey, String refPrimaryLangKey, boolean favorite) {
        this.key = key;
        this.langCode = langCode;
        this.setName(name);
        this.setColor(color);
        this.setIcon(icon);
        this.order = order;
        this.primaryLangKey = primaryLangKey;
        if (refKey != null) {
            this.refKey = refKey;
        }
        if (refPrimaryLangKey != null) {
            this.refPrimaryLangKey = refPrimaryLangKey;
        }
        this.favorite = favorite;
    }

    public Category(int mark, String icon) {
        this("", "", "", "", icon, 0, "");
        this.mark = mark;
    }

    public Category primaryLangCategory;

    public Category getPrimaryLangCategory() {
        if (!TextUtils.isEmpty(primaryLangKey)) {
            if (primaryLangCategory == null) {
                primaryLangCategory = Cache.getInstance().lookupCategory(primaryLangKey);
            }
            return primaryLangCategory;
        }
        return null;
    }

    public String getEffectiveKey() {
        if (refPrimaryLangKey != null) {
            return refPrimaryLangKey;
        }
        return !TextUtils.isEmpty(primaryLangKey) ? primaryLangKey : key;
    }

    public void deriveFrom(Category category) {
        this.color = category.color;
        this.icon = category.icon;
    }

    public JSONObject toJSON() throws JSONException {
        JSONObject json = new JSONObject();
        json.put("k", key);
        json.put("l", langCode);
        json.put("n", name);
        json.put("c", color);
        json.put("i", icon);
        json.put("o", order);
        json.put("p", primaryLangKey);
        json.put("r", refKey != null ? refKey : "");
        json.put("rp", refPrimaryLangKey != null ? refPrimaryLangKey : "");
        json.put("s", space);
        return json;
    }

    public String toJSONString() throws JSONException {
        return toJSON().toString(0);
    }

    public static Category fromJSON(JSONObject json) throws JSONException {
        Category category = new Category(
                json.getString("k"),
                json.getString("l"),
                json.getString("n"),
                json.getString("c"),
                json.getString("i"),
                json.getInt("o"),
                json.getString("p"),
                !TextUtils.isEmpty(json.getString("r")) ? json.getString("r") : null,
                !TextUtils.isEmpty(json.getString("rp")) ? json.getString("rp") : null);
        category.space = json.getString("s");
        return category;
    }

    public static Category fromJSONString(String jsonString) throws JSONException {
        return fromJSON(new JSONObject(jsonString));
    }

    @Override
    public String toString() {
        return name;
    }

    @Override
    public int hashCode() {
        return this.key.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj != null && obj instanceof Category) {
            if (((Category) obj).key.equals(this.key)) {
                return true;
            }
        }
        return false;
    }
}

File: iOS/SayHi/SayHi/Category.swift
//
//  Category.swift
//  SayHi
//
//  Created by Klemenz, Oliver on 13.10.16.
//  Copyright © 2016 Klemenz, Oliver. All rights reserved.
//

import UIKit
import Foundation
import SwiftyJSON

let CategoryStagedIcon = "staged"
let CategoryStagedColor = "#999999"

let CategoryFavorite = Category(mark: 1, icon: "favorite")
let CategorySearch = Category(mark: 2, icon: "search")
let CategoryOwn = Category(mark: 3, icon: "own")
let CategoryStaged = Category(mark: 4, icon: "staged")
let CategoryMore = Category(mark: 5, icon: "more")

class Category: Equatable {
        
    var key: String = ""
    var langCode: String = ""
    var _name: String = ""
    var name: String {
        set {
            _name = newValue
            search = name.searchNormalized(langCode: langCode)
        }
        get {
            return _name
        }
    }
    var _color: String = ""
    var color: String {
        set {
            _color = newValue
            if !color.isEmpty {
                self.bgColor = UIColor.colorWithHexString(hexString: color)
                self.textColor = self.bgColor.isLight() ? UIColor.black : UIColor.white
            } else {
                self.bgColor = UIColor.clear
                self.textColor = UIColor.white
            }
        }
        get {
            return _color
        }
    }
    var _icon: String = ""
    var icon: String {
        set {
            _icon = newValue
            if !self.icon.isEmpty {
                self.iconImage = IconService.instance.icon(self.icon)
            }
        }
        get {
            return _icon
        }
    }
    var order: Int = 0
    
    var primaryLangKey: String = ""
    var refKey: String?
    var _refPrimaryLangKey: String?
    var refPrimaryLangKey: String? {
        set { _refPrimaryLangKey = newValue }
        get { return _refPrimaryLangKey ?? refKey }
    }

    var bgColor: UIColor = UIColor.white
    var textColor: UIColor = UIColor.black
    var iconImage: UIImage?

    var search: String = ""
    var favorite: Bool = false
    var selected: Bool = false
    var stage: Bool = false
    var mark: Int = 0
    
    var space: String = ""
    
    init(key: String, langCode: String, name: String,  color: String, icon: String, order: Int, primaryLangKey: String, refKey: String? = nil, refPrimaryLangKey: String? = nil, favorite: Bool = false) {
        self.key = key
        self.langCode = langCode
        self.name = name
        self.color = color
        self.icon = icon
        self.order = order
        self.primaryLangKey = primaryLangKey
        if let refKey = refKey {
            self.refKey = refKey
        }
        if let refPrimaryLangKey = refPrimaryLangKey {
            self.refPrimaryLangKey = refPrimaryLangKey
        }
        self.favorite = favorite
    }
    
    convenience init(mark: Int, icon: String) {
        self.init(key: "", langCode: "", name: "", color: "", icon: icon, order: 0, primaryLangKey: "")
        self.mark = mark
    }
    
    var _primaryLangCategory: Category?
    var primaryLangCategory: Category? {
        if !primaryLangKey.isEmpty {
            if _primaryLangCategory == nil {
                _primaryLangCategory = Cache.instance.lookupCategory(key: primaryLangKey)
            }
            return _primaryLangCategory
        }
        return nil
    }
    
    var effectiveKey: String {
        get {
            if let refPrimaryLangKey = refPrimaryLangKey {
                return refPrimaryLangKey
            }
            return !primaryLangKey.isEmpty ? primaryLangKey : key
        }
    }
    
    var description: String {
        return name
    }
    
    func deriveFrom(_ category: Category) {
        self.color = category.color
        self.icon = category.icon
    }
    
    func toJSON() -> JSON {
        var raw: [String:Any] = [:]
        raw["k"] = key
        raw["l"] = langCode
        raw["n"] = name
        raw["c"] = color
        raw["i"] = icon
        raw["o"] = order
        raw["p"] = primaryLangKey
        raw["r"] = refKey ?? ""
        raw["rp"] = refPrimaryLangKey ?? ""
        raw["s"] = space
        return JSON(raw)
    }
    
    func toJSONString() -> String {
        return toJSON().rawString(String.Encoding.utf8, options: JSONSerialization.WritingOptions(rawValue: 0))!
    }
    
    static func fromJSON(json : JSON) -> Category {
        var raw = json.rawValue as! [String:Any]
        let category = Category(key: raw["k"] as! String,
                                langCode: raw["l"] as! String,
                                name: raw["n"] as! String,
                                color: raw["c"] as! String,
                                icon: raw["i"] as! String,
                                order: raw["o"] as! Int,
                                primaryLangKey : raw["p"] as! String,
                                refKey: !(raw["r"] as! String).isEmpty ? raw["r"] as? String : nil,
                                refPrimaryLangKey: !(raw["rp"] as! String).isEmpty ? raw["rp"] as? String : nil)
        category.space = raw["s"] as! String
        return category
    }
    
    static func fromJSONString(jsonString : String) -> Category? {
        if let data = jsonString.data(using: String.Encoding.utf8, allowLossyConversion: false) {
            return Category.fromJSON(json: JSON(data: data))
        }
        return nil
    }
}

func ==(lhs: Category, rhs: Category) -> Bool {
    return lhs.key == rhs.key && lhs.mark == rhs.mark
}
